toc/rfc/rfc-draft-bindable-secrets.md

```diff
@@ -0,0 +1,254 @@
+# Meta
+[meta]: #meta
+- Name: Bindable Secrets
+- Start Date: 2024-10-08
+- Author(s): @tcdowney @gerg
+- Status: Draft
+- RFC Pull Request: https://github.com/cloudfoundry/community/pull/994
+
+## Summary
+
+**tl;dr** Kubernetes-style Secrets for Cloud Foundry
+
+This RFC proposes adding a `secret` resource to the V3 Cloud Foundry API, similar in design and purpose to the Kubernetes [Secret resource](https://kubernetes.io/docs/concepts/configuration/secret/). Cloud Foundry Secrets will store arbitrary data in [CredHub](https://docs.cloudfoundry.org/credhub/) inject it into app containers via `tmpfs`-mounted files at a given path, using the same mechanism as [RFC 0030 - Add Support for File based Service Binding Information](https://github.com/cloudfoundry/community/blob/main/toc/rfc/rfc-0030-add-support-for-file-based-service-binding.md).
+
+## Problem
+
+Cloud Foundry currently supports app configuration via environment variables. Other application platforms, such as Kubernetes, support app configuration at runtime [via volume-mounted files](https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod). Many apps are adopting this file-based configuration approach, since environment variables have length limits and files can be updated without having to recreate the container. This RFC proposes introducing file-based configuration to support these sorts of apps on Cloud Foundry — without requiring app code changes.
+
+### What About User-Provided Services?
+
+Why can't developers use user-provided service instances (UPSIs) and service bindings for file-based app configuration? Some apps may have file-based config that don't naturally fit into the service binding model. For example, if an app relies on a config file at a certain path, that plath may not be configurable to `$SERVICE_BINDING_ROOT`. This is especially the case for commercial off-the-shelf software (COTS).
+
+## Proposal
+
+Add two new resources to the V3 Cloud Foundry API: `secret` and the `secret_binding`, along with associated CRUD API endpoints.
+
+### Why Top-Level Resources?
+
+Why not represent secrets as sub-resources of apps like environment variables? Secrets may be shared across multiple apps. For example, a "logical app" may actually be several Cloud Foundry apps resources that need to share certain configuration, certificates, or other credentials. By scoping the `secret` at the `space` level, we can easily support this. Similarly, an app using the blue-green deployment pattern can easily share configuration between versions.
+
+### `secret` and `secret_binding` Resources
+The `secret` will be a `space`-scoped resource (similar to a `route` or a `service_instance`) that can be bound to one or more `app`s in the `space` using a `secret_binding`.
```

Comment:
**question** Both `routes` and `service_instances` are sharable with other spaces.  Do you anticipate 'shared secrets' as well?

Reply:
Heh, I hadn't considered that. I don't think it would be in scope for an MVP, but that is probably something people would want eventually. That's one of the most used features of Carvel's secret-gen-controller on K8s for example. https://github.com/carvel-dev/secretgen-controller/blob/develop/docs/secret-export.md

What is missing from the design to support this? Does sharing require a separate table?

Reply:
I'm not super familiar with the resource sharing code, but I think `routes` and `service_instances` are linked to other spaces via a separate table/model (`RouteShare` and `ServiceInstanceShare`).

Mostly asking as I'm wondering if we should start thinking about making the Sharable modeling or code more generic for if/when we want to make secrets (or future resources) sharable, as it seems to be becoming a pattern.

Reply:
Secrets are not shared with other namespaces in secretgen. What it does is the reconciler will copy these secrets from the source to the other namespaces and ensure that whenever there is a change in the original secret it will get copied to the namespaces where the export is done to.
Is your suggestion to do something similar @tcdowney? 

Another thing that came to mind when I was reading the proposal was the ability to have Organization level secrets, pretty much like github does with their secrets where you can define secrets in the organization and they can be used in all the repositories. Is this something that yll thought about or want to make possible one day?

Reply:
I was just mentioning some prior art. I think this is an interesting and possible future enhancement for Secrets, but I want to keep scope limited for now. If we develop Secrets and this use case comes up we can revisit in a follow-up RFC.